[WIP] typed process dsl

rkuhn opened the pull request:
This is still work in progress, but it is shaping up.

Missing:

* [x] handling ShortCircuit
* [x] all State handlers

ktoso commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -186,7 +186,14 @@ class StubbedActorContext[T](
 
-  def spawnAdapter[U](f: U ⇒ T): ActorRef[U] = spawnAnonymous[Any](Behavior.emptyBehavior)
+  def spawnAdapter[U](f: U ⇒ T): ActorRef[U] = {
+    val i = Inbox[U](childName.next())
```
childName.next effectively looks the same as anonymous child?

rkuhn replied:
They share the same counter, but that should not be a problem. The important difference is that the adapter is actually functional in that it will enqueue to the main inbox (otherwise the synchronous tests would not be possible).

ktoso commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -186,7 +186,14 @@ class StubbedActorContext[T](
+    new internal.FunctionRef[U](
+      self.path / i.ref.path.name,
+      (msg, _) ⇒ { val m = f(msg); if (m != null) { inbox.ref ! m; i.ref ! msg } },
+      (self) ⇒ inbox.ref.sorry.sendSystem(internal.DeathWatchNotification(self, null)))
```
lol, so `sorry` remained in there :rofl:

ktoso commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -199,6 +206,11 @@ class StubbedActorContext[T](
   }
 
   /**
+   * Retrieve the named inbox.
```
"the childs inbox"?

ktoso commented on akka-typed/src/main/scala/akka/typed/ActorRef.scala:
```diff
@@ -9,6 +9,13 @@ import language.implicitConversions
 import scala.concurrent.Future
 
 /**
+ * Helper parent to allow [[ActorRef]] to be used as key in TypedMultiMap.
```
Ah, reminds me of the discussion with Martin some time ago. Looking forward to scroll down and find that map then...

rkuhn replied:
Aeh, nope, this is just left over from some work that has not survived—but I should actually include a test that this comment does indeed hold water.

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+  /*
+   * Convenient access to the core operations, will be even nicer with Dotty implicit function values
+   */
+  sealed trait OpDSL extends Any {
```
`extends Any` to make it an universal trait?

rkuhn replied:
Nope, to make it absolutely impossible to call any methods on it—because the only value that will ever exist is `null`.

ktoso replied:
Ok cool, shall I comment if I find such spots so that we don't miss any chances do to so?

rkuhn replied:
WDYM? Finding other places where to use this trick?

ktoso replied:
Just in case there's more places we can cache functions (in this PR).

rkuhn replied:
Ah, you might have wanted to place this comment further up, but yes, please watch out for that.

ktoso replied:
Ah yeah, added this comment on the wrong line. Damn collapsing diffs....

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+  /**
+   * This implicit expresses that operations that do not use their input channel can be used in any context.
+   */
+  private implicit def nothingIsSomething[T, U](op: Operation[Nothing, T]): Operation[U, T] = op.asInstanceOf[Operation[U, T]]
```
👍

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+object ScalaProcess {
+  import ScalaProcess._
+
+  class RetriesExceeded extends RuntimeException
```
final or hidden?

patriknw replied:
NoStackTrace? and then it could be an object?

rkuhn replied:
That depends on whether the stack trace has a chance to point out where the failure occurred—will have to try.

rkuhn replied:
done

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+  /**
+   * Helper to make `Operation.map` behave like `flatMap` when needed.
+   */
+  object MapAdapter extends MapAdapterLow {
```
Boring but let's keep the `...LowPriorityImplicits` naming (we have it in other places)

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+   *
+   *  - a Process has a 1:1 relationship with an ActorRef
+   *  - an Operation is a step that a Process takes and that produces a value
+   *  - Processes are concurrent, but not distributed: all failures stop the entire Actor
```
So I guess this is aimed to work well with the new supervision scheme you've been envisioning?
If we'd plug persistence, a store would be not per operation but per "successful application of process", so a fialure to apply a process would result in no persist happening? Trying to see how I'd design persistence using these hm...

rkuhn replied:
Supervision is unchanged, but what is important is that allowing partial failure within an Actor makes it internally “distributed”—with all the problems but none of the benefits. Hence any exception that escapes will just kill it (or restart, using the supervision combinator from the ScalaDSL).

ktoso replied:
Yeah, exactly - I meant the "supervision combinators" can be used to get a resume rather easily etc. I agree on the fail fast thing too. If I have "steps" then i can model them as steps of different behaviours I guess? Will get to that part soon in the PR I guess.

(Example being: command(fetch site A, fetch site B), so I want to store the results individually)

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+   */
+  trait MapAdapterLow {
+    implicit def mapAdapterAny[Self, Out]: MapAdapter[Self, Out, Out] =
+      new MapAdapter[Self, Out, Out] {
```
could be cached; not sure if used enough to matter

rkuhn replied:
it should be cached before this PR is merged, but I left it for now for clarity

rkuhn replied:
done

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+   * to `Duration.Inf` for a server process.
+   */
+  final case class Process[S, +Out](name: String, timeout: Duration, mailboxCapacity: Int, operation: Operation[S, Out]) {
+    if (name != "") a.ActorPath.validatePathElement(name)
```
why is empty allowed?

rkuhn replied:
because this is not the full name: keeping those unique would be too much of a burden, and it would not be composable; this name will eventually be appended to the generated `$!a` etc. (but that is not yet implemented)

rkuhn replied:
it is implemented and tested now

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+   */
+
+  private[typed] case class FlatMap[S, Out1, Out2](first: Operation[S, Out1], then: Out1 ⇒ Operation[S, Out2]) extends Operation[S, Out2] {
+    override def toString: String = s"FlatMap($first)"
```
not sure I like hiding the 2nd param. Often confuses people "wat, I thought it had 2 params!".
Perhaps `s"FlatMap($first, <function1:then>)`?

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+
+  /*
+   * The core operations: keep these minimal!
+   */
```
I fear there will have to be some bikeshed about the`op` prefix there... I see why with prefix and not plain, since people will want to call the values `system` for example etc, but perhaps we can get away with something nicer.
E.g. why not keep the UpperCase directly for the ops?

```
for { 
  system <- System
  self <- ActorSelf
] 
```

not too bad?

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+   * filled again using `updateState` or `replayPersistentState`.
+   */
+  def opForgetState[T](key: StateKey[T])(implicit opDSL: OpDSL): Operation[opDSL.Self, akka.Done] =
+    Forget(key)
```
great name 👍

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+    def iterator: Iterator[akka.typed.ScalaProcess.SetStateEvent[T]] = Iterator.single(this)
+    def apply(idx: Int): akka.typed.ScalaProcess.SetStateEvent[T] =
+      if (idx == 0) this
+      else throw new IndexOutOfBoundsException
```
Add message?

rkuhn replied:
right

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+
+  /**
+   * The data for a [[StateKey]] of this kind can be made persistent by
+   * invoking `replayPersistentState`. Persistence is achieved by writing all
```
made persistent by replay? Seems somewhat inverted

rkuhn replied:
right, reworded

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+   * current process to communicate results. The returned [[SubActor]] reference
+   * can be used to send messages to the forked process or to cancel it.
+   */
+  def opFork[Self](process: Process[Self, Any])(implicit opDSL: OpDSL): Operation[opDSL.Self, SubActor[Self]] =
```
Ah, excellent. Was about to comment that "oh but we'll need a new verb because child actors..." but fork already is that new verb. Very nice actually – fork fits the SubActor well as does spawn for children. Thumbs up here.

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+      .map {
+        case Some(res)           ⇒ opUnit(res)
+        case None if retries > 0 ⇒ retry(timeout, retries - 1, ops)
+        case None                ⇒ throw new RetriesExceeded
```
Add message.

rkuhn replied:
done

ktoso commented:
Did not finish reviewing tonight not looking very interesting. Will continue later and want to give it a spin eventually :)

patriknw commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+  private[typed] case object Read extends Operation[Nothing, Nothing]
+  private[typed] case object ProcessSelf extends Operation[Nothing, ActorRef[Any]]
+  private[typed] case object ActorSelf extends Operation[Nothing, ActorRef[ActorCmd[Nothing]]]
+  private[typed] case class Return[T](value: T) extends Operation[Nothing, T]
```
make it a habit to use `final` for all case classes

rkuhn replied:
okay, can certainly do, but thought it was not needed for private ones

patriknw replied:
the pattern matcher takes advantage of it (not sure exactly what the difference is)

patriknw commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+   * await and return that process’ result. This is equivalent to creating
+   * a process with [[OpDSL]] and using `call` to execute it.
+   */
+  def opNextStep[T] =
```
explicit return type

rkuhn replied:
yes

patriknw commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -0,0 +1,538 @@
+   * `afterUpdates` is `true`.
+   */
+  def opUpdateState[T, E](key: StateKey[T], afterUpdates: Boolean = true)(
+    transform: T ⇒ (Seq[key.Event], E))(implicit opDSL: OpDSL): Operation[opDSL.Self, E] =
```
`immutable.Seq` (several places)

rkuhn replied:
yes

patriknw reviewed (commented):
very interesting, looking good

rkuhn commented on akka-typed/src/test/scala/akka/typed/ProcessSpec.scala:
```diff
@@ -0,0 +1,346 @@
+import ScalaProcess._
+import patterns.Receptionist._
+import scala.concurrent.duration._
+import AskPattern._
```
okay

rkuhn reviewed (commented):
Thanks for the reviews @ktoso and @patriknw ! There are indeed several bike sheds that lack proper painting, but I’ll focus on making it work first.

rkuhn commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -199,6 +206,11 @@ class StubbedActorContext[T](
   /**
+   * Retrieve the named inbox.
+   */
+  def getInbox[U](name: String): Inbox[U] = _children(name).asInstanceOf[Inbox[U]]
```
good point, cleaned it up (`selfInbox` and `childInbox` now)

rkuhn committed:
add Process DSL and interpreter, #22105

The point is to allow the creation and reuse of behavior snippets from
which Actor behaviors can be composed.

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -435,10 +436,16 @@ object ScalaProcess {
    * being dispatched—it might still be delivered if it was previously dispatched.
+   *
+   * If `onFailure` is provided it can override the value to be sent if the
+   * watched Actor failed and was a child Actor of the Actor hosting this process.
```
sounds more like `mapWatchMessage` or something like that? We have mapError in streams so being similar would be nice

ktoso commented on akka-typed/src/main/scala/akka/typed/ScalaProcess.scala:
```diff
@@ -435,10 +436,16 @@ object ScalaProcess {
+    opDSL: OpDSL): Operation[opDSL.Self, a.Cancellable] =
+    WatchRef(watchee, msg, target, onFailure)
 
+  private[typed] val any2none = (_: Any) ⇒ None
```
we have this in `ConstantFun` I believe

rkuhn commented:
refs #22105

rkuhn commented:
This PR cannot possibly have broken remoting (at least I would be surprised), seems like a random timeout.

In other news: is this something that can in principle be merged? Should I be working on the reference docs?

ktoso commented:
I think it's going in the right direction though I wanted to re-read the tests again. Ref docs would be great. I had some trouble understanding the exact semantics of the tests actually, need to sit down and "get" it, hope to read through during the weekend again :)

patriknw commented:
PLS BUILD

patriknw commented:
PLS BUILD

rkuhn commented:
The process DSL shall mature at rkuhn/akka-typed-session for a while first, closing this PR. I’ll open a new one with the small modifications needed in the core to make the DSL work as an external package.
